src/include/storage/lock.h

```diff
@@ -178,11 +179,16 @@ typedef struct LOCKTAG
  * rather than accessing the fields directly.  Note multiple eval of target!
  */
 
+/*
+ * For SET_LOCKTAG_RELATION, if it is an ENR temp table we also add MyProcPid into the locktag
+ * in order to disambiguate locks from different backends in case we enter into the shared lock space.
+ */
+
 /* ID info for a relation is DB OID + REL OID; DB OID = 0 if shared */
 #define SET_LOCKTAG_RELATION(locktag,dboid,reloid) \
 	((locktag).locktag_field1 = (dboid), \
 	 (locktag).locktag_field2 = (reloid), \
-	 (locktag).locktag_field3 = 0, \
+	 (locktag).locktag_field3 = (get_ENR_withoid(currentQueryEnv, reloid, ENR_TSQL_TEMP) ? MyProcPid : 0), \
```

Comment:
any way we can move this to a separate macro like above? i imagine this macro is called a lot, so the less we need to call get_ENR_withoid, the better

Reply:
As far as I can tell, all of the places that call this macro are where it is possible for the target relation to be an ENR table, so I don't think it would make any difference whether we check here or in the source function.

Reply:
do we need TSQL check here?

Reply:
To be honest, I am not 100% sure whether we need a sql_dialect check here. Do we intend for ENR temp tables/temp tables on RO to only work with TSQL dialect, or are we considering generalizing the implementation to work with non-babelfish instances as well? 

As is, I don't think that it is unsafe to not have a sql_dialect check here - it will simply add a slight overhead due to the function call, but the function itself is already programmed quite defensively.

Reply:
yes, we want temp on RO to only work on TSQL tables. it's a separate discussion whether we eventually want to support a use case on native APG.

Reply:
hm, it turns out that this change may not actually be necessary. I ran a test where I had two sessions which both exceeded the fast path lock limit (so went into shared locks) without this change, and both sessions were able to freely create, insert, update, etc their own temp tables.